### ISSUE TITLE
MPSGraph prints an ANE cast warning on every training step

### ISSUE THREAD

**Issue by BenjaminDEMAILLE** [IPNP-BIPN/tensorflow-metal-plugin#8]
A plain Keras `model.fit` on the GPU prints this once per step, straight from MetalPerformanceShadersGraph:

```
warning: loc("mps_and"(...MPSGraphUtilities.mm":234:0)): failed: ANE I/O op can only do F16 MemRef <-> F32 Tensor cast
```

Reproduced in a clean virtualenv with `tensorflow==2.20.0` and the package installed, on macOS 26.6, running `tests/installed_test.py`: nine lines during the first epoch, none after.

What is known:

- The training is correct. The loss falls, the sweep is at 0 mismatches, and the same script produces the same answers as the CPU.
- The op named is `mps_and`, a boolean AND, which in that graph comes from the sparse categorical cross entropy rather than from anything this plugin writes by hand.
- The message is MPSGraph reporting that it declined to place something on the Neural Engine and fell back. It is a note about placement, not about numerics.

Why it is worth fixing anyway: it is the first thing a user sees, it repeats, and it looks like an error. Either the graph should be built so MPSGraph does not try that placement, or the warning should be suppressed with a note saying why.

Unknown so far: which tensor MPSGraph is trying to cast, and whether the boolean comes from our graph construction or from TensorFlow's own decomposition of the loss.

**Comment by BenjaminDEMAILLE** [comment]
Found, measured and fixed in 867e573 on #9.

**It is not the loss function.** Probing one op at a time under the plugin: `SparseSoftmaxCrossEntropyWithLogits` 0 warnings, `SoftmaxCrossEntropyWithLogits` 0, `MaxPoolGrad` 0, `Conv2DBackpropInput` 0, `Relu6Grad` 0. `LogicalAnd` 9, `LogicalOr` 9, `LogicalNot` 6. It reached the training script through the `accuracy` metric, which is where a boolean operation appears in an otherwise float graph.

**What it is.** MPSGraph tries to place an op whose operands are boolean onto the Neural Engine, the Neural Engine casts only between F16 and F32, the attempt fails and the failure reports itself. Nothing to do with TensorFlow: reproduced in thirty lines of Objective-C++ against MPSGraph directly, no TensorFlow linked.

That standalone probe is also what ruled out the alternatives:

| spelling, on boolean operands | result |
| --- | --- |
| `logicalANDWithPrimaryTensor:` | 9 warnings |
| `minimumWithPrimaryTensor:` | 9 warnings |
| `bitwiseANDWithPrimaryTensor:` | assertion failure inside MPSGraph, process aborts |
| cast to float32, multiply, cast back | silent |

One more thing the probe showed: the warnings appear only when MPSGraph compiles the graph. A second run in the same process, or a run that hits MPSGraph's on-disk cache, prints nothing. That matches the nine lines landing in the first epoch and never again, and it means anyone trying to reproduce this needs a cold cache.

**The fix.** `LogicalAnd` is a float multiply between two casts, `LogicalOr` a maximum, `LogicalNot` a comparison against zero. A TensorFlow boolean holds 0 or 1, so the product is true exactly when both sides are and the maximum is true exactly when either is. Comparing against zero is exact whatever the representation.

**It costs nothing.** Four million elements, 100 calls after 10 warmups, M4 Max: 0.29 ms for the logical spelling, 0.29 ms for the float one, and the ordering flips between repeats.

Verified against numpy for all three ops, warning count is now 0 where it was 9, 9 and 6, and the sweep is unchanged at 268 verified with 0 mismatches.
